Generate Serialization for NSURLRequest

sheeparegreat opened the pull request:
#### 872f4a3d8f55325cd73ed596cce476a3a8e4f2c1
<pre>
Generate Serialization for NSURLRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=269625">https://bugs.webkit.org/show_bug.cgi?id=269625</a>
<a href="https://rdar.apple.com/123121281">rdar://123121281</a>

Reviewed by Alex Christensen.

Covered partially by existing test: WebKit.LoadNSURLRequestWithMutablePropertiesAndKeys.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::encodeObjectDirectly&lt;NSURLRequest&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSURLRequest&gt;):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm: Added.
(WebKit::CoreIPCNSURLRequest::CoreIPCNSURLRequest):
(WebKit::CoreIPCNSURLRequest::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlist.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCPlistArray::CoreIPCPlistArray):
(WebKit::CoreIPCPlistArray::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm: Added.
(WebKit::CoreIPCPlistDictionary::CoreIPCPlistDictionary):
(WebKit::CoreIPCPlistDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCPlistObject::value const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm: Added.
(WebKit::CoreIPCPlistObject::isPlistType):
(WebKit::valueFromID):
(WebKit::CoreIPCPlistObject::CoreIPCPlistObject):
(WebKit::CoreIPCPlistObject::toID const):
(IPC::ArgumentCoder&lt;UniqueRef&lt;WebKit::PlistValue&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;UniqueRef&lt;WebKit::PlistValue&gt;&gt;::decode):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, SecureCoding)):

Canonical link: <a href="https://commits.webkit.org/281502@main">https://commits.webkit.org/281502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bd65a4f514ad065ad321f4c1ebc077595051b41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59930 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10458 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46941 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48569 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7297 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61960 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51884 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29411 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9125 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9379 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53024 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55243 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9406 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65580 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59175 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3860 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55914 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56064 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3198 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80933 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9002 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35091 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14052 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36173 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37261 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35917 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1167,6 +1168,10 @@ static void runTestCF(const CFHolderForTesting& holderArg)
     runTestCF({ nullptr });
 
+    // NSURLRequest
+    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:url];
```
This should really have a more complicated test.

sheeparegreat commented on Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm:
```diff
@@ -148,4 +191,103 @@ - (NSDictionary *)accumulatedDictionary
 
+- (void)encodeValueOfObjCType:(const char *)type at:(const void *)addr
+{
+    // NSURLRequest decoding calls to configure decode parameters
```
Typo: // NSURLRequest encoding calls to configure the decode parameters

sheeparegreat commented on Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm:
```diff
@@ -148,4 +191,103 @@ - (NSDictionary *)accumulatedDictionary
+
+    // NSURLRequest -initWithCoder calls [coder allowedClasses] and then into decodeObjectOfClass* methods on WKKeyedCoder.
+    // This code is expecting to be called when decoding an NSURLRequest
+    m_handling_NSURLRequest = YES;
```
This should be configured when making the WKKeyedCoder based on the need to decode an NSURLRequest.

sheeparegreat commented on Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm:
```diff
@@ -148,4 +191,103 @@ - (NSDictionary *)accumulatedDictionary
+    // This code is expecting to be called when decoding an NSURLRequest
+    m_handling_NSURLRequest = YES;
+
+    if ([m_dictionary count] < 3) {
```
This should be more explicit and have more tests.

sheeparegreat commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1353,6 +1354,11 @@ - (instancetype)initWithMerchantIdentifier:(NSString *)merchantIdentifier
 
 TEST(IPCSerialization, SecureCoding)
 {
+    // NSURLRequest
```
This should really have a more complicated test.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+#if PLATFORM(COCOA)
+
+[WebKitSecureCodingClass=NSURLRequest.class] webkit_secure_coding NSURLRequest {
+# obj0: isMutable
```
We probably don't really need to send this in perpetuity, but there may be some current bugs that assume mutability is preserved across IPC that we'll need to fix first.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj1: URL
+# obj2: Timeout
+    obj2: Number
+# obj3: Cache Policy
```
This should become an enum.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj4: URL?
+# obj5: Should handle HTTP Cookies
+    obj5: CFBooleanRef
+# obj6: Explicit flags
```
This should become an OptionSet.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj13: Number
+# obj14: Network service type
+    obj14: Number
+# obj15: Request Priority
```
I think this should probably become an enum.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj12: Number
+# obj13: Time window duration
+    obj13: Number
+# obj14: Network service type
```
This should probably become an enum.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+# obj15: Request Priority
+    obj15: Number
+
+# obj17: Cookie policy
```
This should probably become an optional enum

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+
+# obj17: Cookie policy
+    obj17: Number?
+# obj18: Request Method
```
enum, I think

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj19: Dictionary?
+# obj20: Request body data for sending
+    obj20: Data?
+# obj21: Body parts, Array<Data|String>?
```
I'm not sure if we need both of these.  We actually probably need neither as long as EncodeRequestBody exists.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    protocolProperties: Dictionary
+    requiresDNSSECValidation: CFBooleanRef?
+    requiresShortConnectionTimeout: CFBooleanRef
+    schemeWasUpgradedDueToDynamicHSTS: CFBooleanRef
```
I think this is only set in the network process and only used in the network process, so it doesn't need to go across IPC.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+# contentDispositionEncodingFallbackArray: NSArray<NSNumber>
+    contentDispositionEncodingFallbackArray: Array?
+    failInsecureLoadWithHTTPSDNSRecord: CFBooleanRef
+    ignoreHSTS: CFBooleanRef
```
I think this is only set in the network process and only used in the network process, so it doesn't need to go across IPC.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    obj21: Array?
+
+    allowPrivateAccessTokensForThirdParty: CFBooleanRef
+    allowedProtocolTypes: Number
```
Not sure

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    ignoreHSTS: CFBooleanRef
+    isWebSearchContent: CFBooleanRef
+    knownTracker: CFBooleanRef
+    objectCount: Number
```
Almost certainly not needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,100 @@
+    knownTracker: CFBooleanRef
+    objectCount: Number
+    payloadTransmissionTimeout: Number
+    preventHSTSStorage: CFBooleanRef
```
I think this is only set in the network process and only used in the network process, so it doesn't need to go across IPC.

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,116 @@
+    std::optional<CoreIPCNumber> timeout;
+    std::optional<CoreIPCNumber> cachePolicy;
+    std::optional<CoreIPCURL> mainDocumentURL;
+    std::optional<CoreIPCNumber> shouldHandleHTTPCookies;
```
Make these bools

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,116 @@
+namespace WebKit {
+
+struct CoreIPCNSURLRequestData {
+    std::optional<CoreIPCDictionary> protocolProperties;
```
Make specialization for this dict

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,116 @@
+    std::optional<CoreIPCNumber> requestPriority;
+    std::optional<CoreIPCNumber> isHTTP;
+    std::optional<CoreIPCString> httpMethod;
+    std::optional<CoreIPCDictionary> headerFields;
```
Make specialization for this dict

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,156 @@
+    std::optional<NSURLRequestCachePolicy> cachePolicy;
+    std::optional<CoreIPCURL> mainDocumentURL;
+    std::optional<bool> shouldHandleHTTPCookies;
+    std::optional<long> explicitFlags;
```
Make an enumeration for this

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h:
```diff
@@ -0,0 +1,80 @@
+#include <wtf/RetainPtr.h>
+#include <wtf/UniqueRef.h>
+
+
```
remove empty line

sheeparegreat commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
```diff
@@ -361,6 +366,10 @@ WKDDActionContext wrapped by CoreIPCDDSecureActionContext
 #endif
 #endif
 
+#if HAVE(HAVE_WK_SECURE_CODING_NSURLREQUEST)
```
remove HAVE_

sheeparegreat commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
```diff
@@ -109,7 +109,12 @@ header: <WebCore/DictionaryPopupInfo.h>
 
 header: <WebCore/ResourceRequest.h>
 [CustomHeader] struct WebCore::ResourceRequestPlatformData {
+#if HAVE(HAVE_WK_SECURE_CODING_NSURLREQUEST)
```
Fix this

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+
+enum class NSURLRequestNetworkServiceType : uint8_t {
+    Default = 0,
+    OpaqueValue0 = 1,
```
= 0,1,2,3... not needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+};
+
+enum class NSURLRequestAllowedProtocolTypes : uint8_t {
+    OpaqueValue0 = 0,
```
This shouldn't be needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+
+enum class NSURLRequestAllowedProtocolTypes : uint8_t {
+    OpaqueValue0 = 0,
+    OpaqueValue1 = 1,
```
(1 << 0) to match the others?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+
+enum class NSURLRequestAttribution : uint8_t {
+    Developer = 0,
+    User = 1
```
=0, =1 not needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+};
+
+enum class NSURLRequestFlags : int16_t {
+    None = 0,
```
Can we remove none?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+
+enum class NSURLRequestFlags : int16_t {
+    None = 0,
+    ShouldHandleCookies = 2,
```
(1 << 1) like NSURLRequestAllowedProtocolTypes?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+    using HeaderHolder = std::pair<String, std::variant<String, Vector<String>>>;
+
+    std::optional<CoreIPCPlistDictionary> protocolProperties;
+    std::optional<bool> isMutable;
```
This probably shouldn't be optional.  Same with many of these.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,170 @@
+
+struct CoreIPCNSURLRequestData {
+
+    using BodyPartsHolder = std::variant<CoreIPCString, CoreIPCData>;
```
Can we remove "Holder" from the name of this and HeaderHolder?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,329 @@
+
+CoreIPCNSURLRequest::CoreIPCNSURLRequest(CoreIPCNSURLRequestData&& data)
+{
+    m_data = WTFMove(data);
```
nit: 
: m_data(WTFMove(data)) { }

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -1377,6 +1379,26 @@ - (instancetype)initWithMerchantIdentifier:(NSString *)merchantIdentifier
+    [postRequest _setPrivacyProxyFailClosed:YES];
+
+    runTestNS({ postRequest });
+
```
IPCTestingAPI.SerializedTypeInfo probably needs !HAVE(WK_SECURE_CODING_NSURLREQUEST) around @"NSURLRequest" to pass after this.  It's a debug-only API test, so it doesn't get coverage in EWS.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm:
```diff
@@ -0,0 +1,120 @@
+    WTF::switchOn(*m_value, [&](auto& object) {
+        result = object.toID();
+    }, [](std::nullptr_t) {
+        // result should be nil, which is the default value initialized above.
```
This would be slightly nicer like this:

return WTF::switchOn(*m_value, [&](auto& object) {
    return object.toID();
}, [](std::nullptr_t) {
    return RetainPtr<id> { };
});

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm:
```diff
@@ -0,0 +1,120 @@
+
+bool CoreIPCPlistObject::isPlistType(id value)
+{
+    if ([value isKindOfClass:[NSString class]]
```
WebKit is quite inconsistent, but I kind of like dot syntax a little better for things like this:
[value isKindOfClass:NSString.class]

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+};
+
+enum class NSURLRequestAllowedProtocolTypes : uint8_t {
+    OpaqueValue0,
```
I don't think this is needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+struct CoreIPCNSURLRequestData {
+
+    using BodyParts = std::variant<CoreIPCString, CoreIPCData>;
+    using Header = std::pair<String, std::variant<String, Vector<String>>>;
```
HeaderField

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+    using Header = std::pair<String, std::variant<String, Vector<String>>>;
+
+    std::optional<CoreIPCPlistDictionary> protocolProperties;
+    bool isMutable;
```
Let's give all the bools, doubles, enums and ints default values of 0/false to prevent the possibility of using uninitialized memory.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+    OpaqueValue8 = (1 << 7),
+};
+
+enum class NSURLRequestAttribution : uint8_t {
```
: bool since there are two values?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+    id attribution = dict[@"attribution"];
+    if (attribution != nil && [attribution isKindOfClass:[NSNumber class]]) {
+        auto val = [allowedProtocolTypes unsignedCharValue];
+        m_data.attribution = static_cast<NSURLRequestAttribution>(val);
```
For these cases where we cast to an enum, let's make sure it's a valid value first using isValidEnum.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h:
```diff
@@ -0,0 +1,80 @@
+#include "ArgumentCodersCocoa.h"
+#include <wtf/RetainPtr.h>
+#include <wtf/UniqueRef.h>
+
```
extra space.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm:
```diff
@@ -0,0 +1,77 @@
+            continue;
+
+        if (!CoreIPCPlistObject::isPlistType(value))
+            continue;
```
Maybe put an ASSERT_NOT_REACHED on these so we might catch any non-plist contents we would be discarding in release builds.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.mm:
```diff
@@ -0,0 +1,69 @@
+{
+    auto result = adoptNS([[NSMutableArray alloc] initWithCapacity:m_array.size()]);
+    for (auto& object : m_array)
+        [result addObject:object.toID().get()];
```
If toID() ever returns nil we will crash.  Let's use createNSArray from VectorCocoa.h with a lambda instead.  That has the null check.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm:
```diff
@@ -0,0 +1,77 @@
+{
+    auto result = adoptNS([[NSMutableDictionary alloc] initWithCapacity:m_keyValuePairs.size()]);
+    for (auto& keyValuePair : m_keyValuePairs)
+        [result setObject:keyValuePair.value.toID().get() forKey:keyValuePair.key.toID().get()];
```
Let's null check the key and value RetainPtrs to be a little more robust.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+        [dict setObject: mutArray forKey:@"contentDispositionEncodingFallbackArray"];
+    }
+
+    return adoptNS([[NSURLRequest alloc] _initWithWebKitPropertyListData: dict]);
```
no space after the colon here and a few other places in this file.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+    SET_DICT_FROM_PRIMITIVE(isWebSearchContent, NSInteger, Bool);
+
+    if (m_data.contentDispositionEncodingFallbackArray) {
+        NSMutableArray *mutArray = [NSMutableArray array];
```
Either name it mutableArray or array.
Also, let's use initWithCapacity and adopt it to not fill up the autoreleasepool and to reduce allocations.  Here and a few other places in this file.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+{
+    RELEASE_ASSERT([[NSURLRequest class] instancesRespondToSelector:@selector(_initWithWebKitPropertyListData:)]);
+
+    NSMutableDictionary *dict = [[NSMutableDictionary alloc] initWithCapacity:19];
```
This needs to be adopted or it's a memory leak.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+        NSMutableArray *mutArray = [NSMutableArray array];
+        for (auto& value : *m_data.bodyParts) {
+            WTF::switchOn(value,
+                [&] (const CoreIPCData d) {
```
If you use auto& then these two lambdas can be reduced to one.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+        m_data.NAME = WTFMove(elided_NAME);                  \
+    }
+
+#define MAKE_SEL_NAME(x, y) x ## y
```
SELECTOR
This could also be removed and just use ## inline.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+
+#define SET_NSURLREQUESTDATA(NAME, CLASS, WRAPPER)           \
+    id NAME = dict[@#NAME];                                  \
+    if (NAME != nil && [NAME isKindOfClass:[CLASS class]]) { \
```
I don't think this nil check is necessary.  Would CLASS.class work instead of [CLASS class]?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+#define SET_NSURLREQUESTDATA(NAME, CLASS, WRAPPER)           \
+    id NAME = dict[@#NAME];                                  \
+    if (NAME != nil && [NAME isKindOfClass:[CLASS class]]) { \
+        auto elided_NAME = WRAPPER(NAME);                    \
```
I don't see why "elided" is used here.  I would just name the variable data or member or something.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+
+#define SET_NSURLREQUESTDATA_PRIMITIVE(NAME, CLASS, PRIMITIVE) \
+    id NAME = dict[@#NAME];                                    \
+    if (NAME != nil && [NAME isKindOfClass:[CLASS class]])     \
```
nil check is redundant.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,324 @@
+
+CoreIPCNSURLRequest::CoreIPCNSURLRequest(NSURLRequest *request)
+{
+    RELEASE_ASSERT([request respondsToSelector:@selector(_webKitPropertyListData)]);
```
I think this assertion is redundant.  If it doesn't respond to the selector, the next line will generate an equally useful crash log.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+        RetainPtr<NSMutableDictionary> headerFields = adoptNS([[NSMutableDictionary alloc] initWithCapacity:(*m_data.headerFields).size()]);
+        for (auto& headerPair : *m_data.headerFields) {
+            WTF::switchOn(headerPair.second,
+                [&] (const String s) {
```
const String&

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+                [&] (const String s) {
+                    [headerFields setObject:(NSString *)s forKey:(NSString *)headerPair.first];
+                },
+                [&] (const Vector<String> a) {
```
(const Vector<String>& vector)
Prevents unnecessary copies and has a better variable name.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+    std::optional<CoreIPCData> body;
+    std::optional<Vector<BodyParts>> bodyParts;
+
+    double startTimeoutTime = 0;
```
Let's use { 0 } instead of = 0 for all initializers.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+    double startTimeoutTime = 0;
+    bool requiresShortConnectionTimeout = false;
+    double payloadTransmissionTimeout = 0;
+    NSURLRequestAllowedProtocolTypes allowedProtocolTypes = NSURLRequestAllowedProtocolTypes::Auto;
```
I think this should be an OptionSet<NSURLRequestAllowedProtocolTypes>

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+};
+
+enum class NSURLRequestAllowedProtocolTypes : uint8_t {
+    Auto,
```
This should be removed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in:
```diff
@@ -0,0 +1,151 @@
+}
+
+header: "CoreIPCNSURLRequest.h"
+[CustomHeader, WebKitPlatform] enum class WebKit::NSURLRequestAllowedProtocolTypes : uint8_t {
```
I think this needs [OptionSet]

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+    bool isMutable = false;
+    CoreIPCURL URL; // NOLINT(readability/naming/acronym)
+    double timeout = 0;
+    NSURLRequestCachePolicy cachePolicy;
```
This needs a default value to prevent uninitialized memory.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h:
```diff
@@ -0,0 +1,166 @@
+
+    std::optional<CoreIPCPlistDictionary> protocolProperties;
+    bool isMutable = false;
+    CoreIPCURL URL; // NOLINT(readability/naming/acronym)
```
Can we name this url and remove the NOLINT?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+    SET_NSURLREQUESTDATA_PRIMITIVE(timeout, NSNumber, double);
+
+    id cachePolicy = dict[@"cachePolicy"];
+    if (cachePolicy != nil && [cachePolicy isKindOfClass:[NSNumber class]]) {
```
nil check unneeded because isKindOfClass returns false when called on nil.  Several times in this function.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+    if (allowedProtocolTypes != nil && [allowedProtocolTypes isKindOfClass:[NSNumber class]]) {
+        auto val = [allowedProtocolTypes unsignedCharValue];
+        if (isValidEnum<NSURLRequestAllowedProtocolTypes>(val))
+            m_data.allowedProtocolTypes = static_cast<NSURLRequestAllowedProtocolTypes>(val);
```
This should be OptionSet::fromRaw instead.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+        for (id element in contentDispositionEncodingFallbackArray) {
+            if ([element isKindOfClass:[NSNumber class]]) {
+                CoreIPCNumber tooAdd(element);
+                vector.append(WTFMove(tooAdd));
```
I think there's an extra o in too
This could be a one liner:
vector.append(CoreIPCNumber(element))

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,321 @@
+
+RetainPtr<id> CoreIPCNSURLRequest::toID() const
+{
+    RetainPtr<NSMutableDictionary> dict = adoptNS([[NSMutableDictionary alloc] initWithCapacity:19]);
```
nit: I think the deduction guides on RetainPtr make it so you don't need <NSMutableDictionary> a few times in this function.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h:
```diff
@@ -0,0 +1,79 @@
+class CoreIPCData;
+
+using PlistValue = std::variant<
+    std::nullptr_t,
```
Do we need this?  I don't think you can have null in a plist.

achristensen07 reviewed (changes requested):
I didn't mean to remove all the NSURLRequestAllowedProtocolTypes values, just the 0 value.  That's not needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,328 @@
+
+RetainPtr<id> CoreIPCNSURLRequest::toID() const
+{
+    auto dict = adoptNS([[NSMutableDictionary alloc] initWithCapacity:19]);
```
Where does 19 come from?  There are 43 possible members.  Do they commonly have 19 or fewer?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,328 @@
+                    [headerFields setObject:(NSString *)s forKey:(NSString *)headerPair.first];
+                },
+                [&] (const Vector<String>& vector) {
+                    auto array = adoptNS([[NSMutableArray alloc] initWithCapacity:vector.size()]);
```
This could use makeArray.

achristensen07 commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1423,6 +1423,12 @@
 
+#if ((PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 150000) \
+    || ((PLATFORM(IOS) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 180000) \
+    || (PLATFORM(VISION) && __VISION_OS_VERSION_MIN_REQUIRED >= 20000))
```
We should probably follow up and do this on watch and tv as well.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm:
```diff
@@ -0,0 +1,328 @@
+    SET_DICT_FROM_PRIMITIVE(isWebSearchContent, NSInteger, Bool);
+
+    if (m_data.contentDispositionEncodingFallbackArray) {
+        auto array = adoptNS([[NSMutableArray alloc] initWithCapacity:(*m_data.contentDispositionEncodingFallbackArray).size()]);
```
This could use -> instead of (*).

sheeparegreat committed:
Generate Serialization for NSURLRequest
https://bugs.webkit.org/show_bug.cgi?id=269625
rdar://123121281

Reviewed by Alex Christensen.

Covered partially by existing test: WebKit.LoadNSURLRequestWithMutablePropertiesAndKeys.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSObject<NSSecureCoding>>):
(IPC::encodeObjectDirectly<NSURLRequest>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSURLRequest>):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.h: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.mm: Added.
(WebKit::CoreIPCNSURLRequest::CoreIPCNSURLRequest):
(WebKit::CoreIPCNSURLRequest::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSURLRequest.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlist.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCPlistArray::CoreIPCPlistArray):
(WebKit::CoreIPCPlistArray::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h.
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm: Added.
(WebKit::CoreIPCPlistDictionary::CoreIPCPlistDictionary):
(WebKit::CoreIPCPlistDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCArray.h.
(WebKit::CoreIPCPlistObject::value const):
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm: Added.
(WebKit::CoreIPCPlistObject::isPlistType):
(WebKit::valueFromID):
(WebKit::CoreIPCPlistObject::CoreIPCPlistObject):
(WebKit::CoreIPCPlistObject::toID const):
(IPC::ArgumentCoder<UniqueRef<WebKit::PlistValue>>::encode):
(IPC::ArgumentCoder<UniqueRef<WebKit::PlistValue>>::decode):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST(IPCSerialization, SecureCoding)):

Canonical link: https://commits.webkit.org/281502@main
